### PR TITLE
include type reference directives in .d.ts. files

### PR DESCRIPTION
**Pull request by vladima**
in this PR type reference directives are included in every file that exposes something from type declaration file on its public surface

### CODE REVIEW

File: tests/baselines/reference/typeReferenceDirectives10.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+export interface $ { x }
+
+//// [index.d.ts]
+declare let $: { x: number }
+
+//// [app.ts]
+/// <reference types="lib"/>
+import {$} from "./ref";
+export interface A {
+    x: $
```
Comment (mhegazy) [inline tests/baselines/reference/typeReferenceDirectives10.js]: can you add another test, where `$` is used as a value, as that would resolve to the `let $` and should  the `lib`.

Comment (vladima) [inline tests/baselines/reference/typeReferenceDirectives10.js]: done, any other comments?

**Comment by mhegazy** [comment]
:+1:. looks good. any news about perf?

**Comment by RyanCavanaugh** [comment]
:+1:

**Comment by vladima** [comment]
@mhegazy: these are numbers I'm getting when compiling Angular2: 
```
// master
Parse:	Min: 1.680000	Max: 1.770000	Avg: 1.716000	Var: 0.001344	StdDev: 0.036661
Bind:	Min: 0.990000	Max: 1.020000	Avg: 1.008000	Var: 0.000096	StdDev: 0.009798
Check:	Min: 6.560000	Max: 6.830000	Avg: 6.684000	Var: 0.010024	StdDev: 0.100120
Emit:	Min: 2.570000	Max: 2.660000	Avg: 2.618000	Var: 0.001336	StdDev: 0.036551
Total:	Min: 11.820000	Max: 12.230000	Avg: 12.026000	Var: 0.022904	StdDev: 0.151341
```
Before testing current branch I've added `<reference types="zone.js"/>` to `ng_zone_impl.ts`.
```
Parse:	Min: 1.660000	Max: 1.770000	Avg: 1.714000	Var: 0.001304	StdDev: 0.036111
Bind:	Min: 0.930000	Max: 0.990000	Avg: 0.972000	Var: 0.000496	StdDev: 0.022271
Check:	Min: 6.580000	Max: 6.900000	Avg: 6.742000	Var: 0.010496	StdDev: 0.102450
Emit:	Min: 2.640000	Max: 2.730000	Avg: 2.686000	Var: 0.000904	StdDev: 0.030067
Total:	Min: 11.890000	Max: 12.260000	Avg: 12.116000	Var: 0.015784	StdDev: 0.125634```

I do see minor increase in emit time (~0.07 s) as a cost of checking if symbol originates from the type declaration file.

**Comment by mhegazy** [comment]
looks reasonable. thanks! :+1:

**Comment by mhegazy** [comment]
one remaining piece is the ones defined in tsconfig.json under `types`. how should would these be handled?

**Comment by vladima** [comment]
@mhegazy currently we don;t distinguish if type declaration file was added via `/// <reference types="".."/>` or through tsconfig and always add type reference directive if public surface of the file uses something from this type declaration file.

**Pull request merged**